<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>MaybeOnErrorXTest.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">RxJava_2_x$Unnamed.exec</a> &gt; <a href="index.source.html" class="el_package">io.reactivex.internal.operators.maybe</a> &gt; <span class="el_source">MaybeOnErrorXTest.java</span></div><h1>MaybeOnErrorXTest.java</h1><pre class="source lang-java linenums">/**
 * Copyright (c) 2016-present, RxJava Contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an &quot;AS IS&quot; BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
 * the License for the specific language governing permissions and limitations under the License.
 */

package io.reactivex.internal.operators.maybe;

import java.io.IOException;

import org.junit.Test;

import io.reactivex.*;
import io.reactivex.exceptions.TestException;
import io.reactivex.functions.*;
import io.reactivex.internal.functions.Functions;
import io.reactivex.processors.PublishProcessor;

<span class="fc" id="L26">public class MaybeOnErrorXTest {</span>

    @Test
    public void onErrorReturnConst() {
<span class="fc" id="L30">        Maybe.error(new TestException())</span>
<span class="fc" id="L31">        .onErrorReturnItem(1)</span>
<span class="fc" id="L32">        .test()</span>
<span class="fc" id="L33">        .assertResult(1);</span>
<span class="fc" id="L34">    }</span>

    @Test
    public void onErrorReturn() {
<span class="fc" id="L38">        Maybe.error(new TestException())</span>
<span class="fc" id="L39">        .onErrorReturn(Functions.justFunction(1))</span>
<span class="fc" id="L40">        .test()</span>
<span class="fc" id="L41">        .assertResult(1);</span>
<span class="fc" id="L42">    }</span>

    @Test
    public void onErrorComplete() {
<span class="fc" id="L46">        Maybe.error(new TestException())</span>
<span class="fc" id="L47">        .onErrorComplete()</span>
<span class="fc" id="L48">        .test()</span>
<span class="fc" id="L49">        .assertResult();</span>
<span class="fc" id="L50">    }</span>

    @Test
    public void onErrorCompleteTrue() {
<span class="fc" id="L54">        Maybe.error(new TestException())</span>
<span class="fc" id="L55">        .onErrorComplete(Functions.alwaysTrue())</span>
<span class="fc" id="L56">        .test()</span>
<span class="fc" id="L57">        .assertResult();</span>
<span class="fc" id="L58">    }</span>

    @Test
    public void onErrorCompleteFalse() {
<span class="fc" id="L62">        Maybe.error(new TestException())</span>
<span class="fc" id="L63">        .onErrorComplete(Functions.alwaysFalse())</span>
<span class="fc" id="L64">        .test()</span>
<span class="fc" id="L65">        .assertFailure(TestException.class);</span>
<span class="fc" id="L66">    }</span>

    @SuppressWarnings(&quot;unchecked&quot;)
    @Test
    public void onErrorReturnFunctionThrows() {
<span class="fc" id="L71">        TestHelper.assertCompositeExceptions(Maybe.error(new TestException())</span>
<span class="fc" id="L72">        .onErrorReturn(new Function&lt;Throwable, Object&gt;() {</span>
            @Override
            public Object apply(Throwable v) throws Exception {
<span class="fc" id="L75">                throw new IOException();</span>
            }
        })
<span class="fc" id="L78">        .test(), TestException.class, IOException.class);</span>
<span class="fc" id="L79">    }</span>

    @SuppressWarnings(&quot;unchecked&quot;)
    @Test
    public void onErrorCompletePredicateThrows() {
<span class="fc" id="L84">        TestHelper.assertCompositeExceptions(Maybe.error(new TestException())</span>
<span class="fc" id="L85">        .onErrorComplete(new Predicate&lt;Throwable&gt;() {</span>
            @Override
            public boolean test(Throwable v) throws Exception {
<span class="fc" id="L88">                throw new IOException();</span>
            }
        })
<span class="fc" id="L91">        .test(), TestException.class, IOException.class);</span>
<span class="fc" id="L92">    }</span>

    @Test
    public void onErrorResumeNext() {
<span class="fc" id="L96">        Maybe.error(new TestException())</span>
<span class="fc" id="L97">        .onErrorResumeNext(Functions.justFunction(Maybe.just(1)))</span>
<span class="fc" id="L98">        .test()</span>
<span class="fc" id="L99">        .assertResult(1);</span>
<span class="fc" id="L100">    }</span>

    @Test
    public void onExceptionResumeNext() {
<span class="fc" id="L104">        Maybe.error(new TestException())</span>
<span class="fc" id="L105">        .onExceptionResumeNext(Maybe.just(1))</span>
<span class="fc" id="L106">        .test()</span>
<span class="fc" id="L107">        .assertResult(1);</span>
<span class="fc" id="L108">    }</span>

    @Test
    public void onExceptionResumeNextPassthrough() {
<span class="fc" id="L112">        Maybe.error(new AssertionError())</span>
<span class="fc" id="L113">        .onExceptionResumeNext(Maybe.just(1))</span>
<span class="fc" id="L114">        .test()</span>
<span class="fc" id="L115">        .assertFailure(AssertionError.class);</span>
<span class="fc" id="L116">    }</span>

    @SuppressWarnings(&quot;unchecked&quot;)
    @Test
    public void onErrorResumeNextFunctionThrows() {
<span class="fc" id="L121">        TestHelper.assertCompositeExceptions(Maybe.error(new TestException())</span>
<span class="fc" id="L122">        .onErrorResumeNext(new Function&lt;Throwable, Maybe&lt;Object&gt;&gt;() {</span>
            @Override
            public Maybe&lt;Object&gt; apply(Throwable v) throws Exception {
<span class="fc" id="L125">                throw new IOException();</span>
            }
        })
<span class="fc" id="L128">        .test(), TestException.class, IOException.class);</span>
<span class="fc" id="L129">    }</span>

    @Test
    public void onErrorReturnSuccess() {
<span class="fc" id="L133">        Maybe.just(1)</span>
<span class="fc" id="L134">        .onErrorReturnItem(2)</span>
<span class="fc" id="L135">        .test()</span>
<span class="fc" id="L136">        .assertResult(1);</span>
<span class="fc" id="L137">    }</span>

    @Test
    public void onErrorReturnEmpty() {
<span class="fc" id="L141">        Maybe.&lt;Integer&gt;empty()</span>
<span class="fc" id="L142">        .onErrorReturnItem(2)</span>
<span class="fc" id="L143">        .test()</span>
<span class="fc" id="L144">        .assertResult();</span>
<span class="fc" id="L145">    }</span>

    @Test
    public void onErrorReturnDispose() {
<span class="fc" id="L149">        TestHelper.checkDisposed(PublishProcessor.create().singleElement().onErrorReturnItem(1));</span>
<span class="fc" id="L150">    }</span>

    @Test
    public void onErrorReturnDoubleOnSubscribe() {
<span class="fc" id="L154">        TestHelper.checkDoubleOnSubscribeMaybe(new Function&lt;Maybe&lt;Object&gt;, MaybeSource&lt;Object&gt;&gt;() {</span>
            @Override
            public MaybeSource&lt;Object&gt; apply(Maybe&lt;Object&gt; v) throws Exception {
<span class="fc" id="L157">                return v.onErrorReturnItem(1);</span>
            }
        });
<span class="fc" id="L160">    }</span>

    @Test
    public void onErrorCompleteSuccess() {
<span class="fc" id="L164">        Maybe.just(1)</span>
<span class="fc" id="L165">        .onErrorComplete()</span>
<span class="fc" id="L166">        .test()</span>
<span class="fc" id="L167">        .assertResult(1);</span>
<span class="fc" id="L168">    }</span>

    @Test
    public void onErrorCompleteEmpty() {
<span class="fc" id="L172">        Maybe.&lt;Integer&gt;empty()</span>
<span class="fc" id="L173">        .onErrorComplete()</span>
<span class="fc" id="L174">        .test()</span>
<span class="fc" id="L175">        .assertResult();</span>
<span class="fc" id="L176">    }</span>

    @Test
    public void onErrorCompleteDispose() {
<span class="fc" id="L180">        TestHelper.checkDisposed(PublishProcessor.create().singleElement().onErrorComplete());</span>
<span class="fc" id="L181">    }</span>

    @Test
    public void onErrorCompleteDoubleOnSubscribe() {
<span class="fc" id="L185">        TestHelper.checkDoubleOnSubscribeMaybe(new Function&lt;Maybe&lt;Object&gt;, MaybeSource&lt;Object&gt;&gt;() {</span>
            @Override
            public MaybeSource&lt;Object&gt; apply(Maybe&lt;Object&gt; v) throws Exception {
<span class="fc" id="L188">                return v.onErrorComplete();</span>
            }
        });
<span class="fc" id="L191">    }</span>

    @Test
    public void onErrorNextDispose() {
<span class="fc" id="L195">        TestHelper.checkDisposed(PublishProcessor.create().singleElement().onErrorResumeNext(Maybe.just(1)));</span>
<span class="fc" id="L196">    }</span>

    @Test
    public void onErrorNextDoubleOnSubscribe() {
<span class="fc" id="L200">        TestHelper.checkDoubleOnSubscribeMaybe(new Function&lt;Maybe&lt;Object&gt;, MaybeSource&lt;Object&gt;&gt;() {</span>
            @Override
            public MaybeSource&lt;Object&gt; apply(Maybe&lt;Object&gt; v) throws Exception {
<span class="fc" id="L203">                return v.onErrorResumeNext(Maybe.just(1));</span>
            }
        });
<span class="fc" id="L206">    }</span>

    @Test
    public void onErrorNextIsAlsoError() {
<span class="fc" id="L210">        Maybe.error(new TestException(&quot;Main&quot;))</span>
<span class="fc" id="L211">        .onErrorResumeNext(Maybe.error(new TestException(&quot;Secondary&quot;)))</span>
<span class="fc" id="L212">        .test()</span>
<span class="fc" id="L213">        .assertFailureAndMessage(TestException.class, &quot;Secondary&quot;);</span>
<span class="fc" id="L214">    }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.2.201808211720</span></div></body></html>